sql+pgwire: report recoverable panics as errors to clients.

knz opened the pull request:
Panic recovery is enabled iff the cluster setting
`sql.internal_error_recovery.enabled` is true (it is set to true by
default).

A panic is considered recoverable if it was originally emitted by the `sql` package itself or one of its sub-packages. For example, if the panic originates in the `client` package (e.g. as called by client.Txn() from `sql`), the panic is not considered to be recoverable.

The panic details are logged on the server's logs (and stderr); the client only receives a generic message "internal server error during SQL execution" that invites them to investigate the server (or talk to the DBA).

Fixes #9884.

cc @bdarnell

knz committed:
sql/pgwire: recover from SQL panics, log and close the client connection.

Panic recovery is enabled iff the cluster setting
`sql.internal_error_recovery.enabled` is true (it is set to true by
default).

A panic is considered recoverable if it was originally emitted by the
`sql` package itself or one of its sub-packages. For example, if the
panic originates in the `client` package (e.g. as called by
client.Txn() from `sql`), the panic is not considered to be
recoverable.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15063)
<!-- Reviewable:end -->

knz commented:
The crux of the change is in the 2nd commit. The rest is just plumbing.

andreimatei commented:
Review status: 0 of 169 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/sql/pgwire/pgwire_test.go, line 1702 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2tisnPxDzuqZ9hJuT:-Ki2tisnPxDzuqZ9hJuU:b-55n8xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/pgwire_test.go#L1702)):*
> ```Go
> 	// Check that the server recovers from SQL panics and propagates a proper
> 	// error code to the client.
> 	if _, err := db.Exec(`SELECT crdb_internal.internal_error('mytest')`); err != nil {
> ```

let's also please test somehow to your recovery is not broader than you intended - e.g. 

---

*[pkg/sql/pgwire/v3.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2wFWM4s6aYCInAliT:-Ki2wFWM4s6aYCInAliU:b-meqxr0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L189)):*
> ```Go
> 	sqlMemoryPool *mon.MemoryMonitor
> 
> 	// abortErr, if non-nil at the start of an iteration of the
> ```

does `v3Conn` need to be involved in this panic recovery business? I think it might be cleaner if we can keep it all in the `Executor`; that way `v3Conn` wouldn't need to understand internals of the `sql` package, and wouldn't need to modify its main loop with this awkward `abortErr`. What do you think?

---

*[pkg/sql/pgwire/v3.go, line 1167 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2rhdig9dfvnpmHW5y:-Ki2rhdig9dfvnpmHW5z:b8t0jj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1167)):*
> ```Go
> }
> 
> func (c *v3Conn) captureRecoverablePanic(skip int, err *string) {
> ```

needs some comments. What's `skip`? Why is `err` a pointer and not a return val?

---

*[pkg/sql/pgwire/v3.go, line 1174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2rxpcttb-08pXCfwa:-Ki2rxpcttb-08pXCfwb:b1sgo7l) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1174)):*
> ```Go
> 
> 		log.ReportPanic(c.session.Ctx(), r)
> 		*err = fmt.Sprintf("%v\n\n%s\n", r, debug.Stack())
> ```

don't you want to use `skip` here to trim some of the stack?

---

*[pkg/sql/pgwire/v3.go, line 1210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2vaI9z-bN_rrBCJZA:-Ki2vaIAZkhLSUGIbf6s:bqzrqy9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1210)):*
> ```Go
> 		}
> 		if !strings.HasPrefix(function,
> 			"github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements") {
> ```

If possible, I think it'd be good to have a method with a dedicated name, like `recoverFromPanic(closure)`, to look for.

---

*[pkg/sql/pgwire/v3.go, line 1224 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2ufgf-lgYZ-tXZtCJ:-Ki2ufgf-lgYZ-tXZtCK:b-9q9hld) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1224)):*
> ```Go
> 		f, more := frames.Next()
> 		function := f.Function
> 		if strings.HasPrefix(function, "runtime.") {
> ```

what if we add more `recovers` in the sql package? I think you should look for the deepest `runtime` line in the stack, not just the first group of "runtime" lines. No?

---

*[pkg/sql/pgwire/v3.go, line 1230 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2uHJ1FD6ONrvhOH3E:-Ki2uHJ1FD6ONrvhOH3F:b-l6olvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1230)):*
> ```Go
> 			continue
> 		}
> 		if strings.HasPrefix(function, "github.com/cockroachdb/cockroach/pkg/sql") {
> ```

I think we should be more specific here and and exclude panics from the Executor itself (like those pesky inconsistent txn state ones). I have no confidence that recovering after those is and will stay safe (assuming that there even is such a category of "safe" panics).
Can you look again for the `ExecuteStatements` method and only recover from panics under that?

---

*[pkg/util/stop/stopper.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2r4_hFO7PNAwruLSQ:-Ki2r4_hFO7PNAwruLSR:bzfq03a) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee3919d290e3bc9810b6e854f8495e9a4fbeb95f/pkg/util/stop/stopper.go#L205)):*
> ```Go
> // RunWorker runs the supplied function as a "worker" to be stopped
> // by the stopper. The function <f> is run in a goroutine.
> func (s *Stopper) RunWorker(ctx context.Context, f func()) {
> ```

I think this deserves a comment that `ctx` is not automatically used for running `f`. In depending on how my next comment goes, it might deserve even more comment about what it's used for.

---

*[pkg/util/stop/stopper.go, line 208 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2ost5C4enNmXxtXE3:-Ki2ost5C4enNmXxtXE4:b-3prndd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee3919d290e3bc9810b6e854f8495e9a4fbeb95f/pkg/util/stop/stopper.go#L208)):*
> ```Go
> 	s.stop.Add(1)
> 	go func() {
> 		defer s.Recover(ctx)
> ```

mmm is this safe wrt tracing spans that might have been closed in the meantime? Like, have you looked at all the callers and checked that the span outlives the tasks that they start? I'm sure that in many situations it doesn't. I think you should derive a context with no tracing stuff in it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

knz commented:
@tamird please hold off reviewing the first commit, I'll fork it in a separate PR first.

knz commented:
Review status: 0 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/pgwire/pgwire_test.go, line 1702 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2tisnPxDzuqZ9hJuT:-Ki5E4dY3j_8qHBAvT5Y:bhvkkkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/pgwire_test.go#L1702)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

let's also please test somehow to your recovery is not broader than you intended - e.g. 
</blockquote></details>

Can you make a suggestion? I would like to introduce an artificial panic in the `client` package, but I do not know how to do that.

---

*[pkg/sql/pgwire/v3.go, line 189 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2wFWM4s6aYCInAliT:-Ki5DmXqu-fMJO9KF6ZF:b79hzyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L189)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

does `v3Conn` need to be involved in this panic recovery business? I think it might be cleaner if we can keep it all in the `Executor`; that way `v3Conn` wouldn't need to understand internals of the `sql` package, and wouldn't need to modify its main loop with this awkward `abortErr`. What do you think?
</blockquote></details>

It is good to move most of this logic to Executor. Done.

---

*[pkg/sql/pgwire/v3.go, line 1167 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2rhdig9dfvnpmHW5y:-Ki5CsKQax6bDP0amEcx:b30wvqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1167)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

needs some comments. What's `skip`? Why is `err` a pointer and not a return val?
</blockquote></details>

Function is deleted and merged into ExecuteStatements.

---

*[pkg/sql/pgwire/v3.go, line 1174 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2rxpcttb-08pXCfwa:-Ki5D2fpfV_FJfDT5jZJ:b-vsz1ji) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1174)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

don't you want to use `skip` here to trim some of the stack?
</blockquote></details>

I'm not sure how many levels though. Apparently depending on the system there is 1 or 2 level introduced in the call stack per panic recovery stage.

---

*[pkg/sql/pgwire/v3.go, line 1210 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2vaI9z-bN_rrBCJZA:-Ki5DD6Gz1_sl0zMY8R2:b-tk2pws) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1210)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

If possible, I think it'd be good to have a method with a dedicated name, like `recoverFromPanic(closure)`, to look for.
</blockquote></details>

How do you propose to make this work? Can you provide an example?

---

*[pkg/sql/pgwire/v3.go, line 1224 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2ufgf-lgYZ-tXZtCJ:-Ki5DGKmV0hiJi4DoSvA:b-59oxxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1224)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what if we add more `recovers` in the sql package? I think you should look for the deepest `runtime` line in the stack, not just the first group of "runtime" lines. No?
</blockquote></details>

Yes, this is a good idea. Will do.

---

*[pkg/sql/pgwire/v3.go, line 1230 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2uHJ1FD6ONrvhOH3E:-Ki5DKutOrR76U0g5t-S:bot6bb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9d6e6dfee2d49213b2f20e273625f73c5f2a244d/pkg/sql/pgwire/v3.go#L1230)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think we should be more specific here and and exclude panics from the Executor itself (like those pesky inconsistent txn state ones). I have no confidence that recovering after those is and will stay safe (assuming that there even is such a category of "safe" panics).
Can you look again for the `ExecuteStatements` method and only recover from panics under that?
</blockquote></details>

I don't understand? Everything is under ExecuteStatements, because that's the top-level function called by pgwire.
Perhaps you mean `runTxnAttempt`?

---

*[pkg/util/stop/stopper.go, line 205 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2r4_hFO7PNAwruLSQ:-Ki58V5859m2H2y6ua-S:b-srzkm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee3919d290e3bc9810b6e854f8495e9a4fbeb95f/pkg/util/stop/stopper.go#L205)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I think this deserves a comment that `ctx` is not automatically used for running `f`. In depending on how my next comment goes, it might deserve even more comment about what it's used for.
</blockquote></details>

Moved off to #15153.

---

*[pkg/util/stop/stopper.go, line 208 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Ki2ost5C4enNmXxtXE3:-Ki58ZiIETNUdfG2Z0zj:b-zf2xye) ([raw file](https://github.com/cockroachdb/cockroach/blob/ee3919d290e3bc9810b6e854f8495e9a4fbeb95f/pkg/util/stop/stopper.go#L208)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mmm is this safe wrt tracing spans that might have been closed in the meantime? Like, have you looked at all the callers and checked that the span outlives the tasks that they start? I'm sure that in many situations it doesn't. I think you should derive a context with no tracing stuff in it.
</blockquote></details>

Yes, I have looked at all the callers. It was painful, and I may have made mistakes, but overall I think we're good.
Anyway this is now separate - moved off to #15153..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I can think of two possible problems with swallowing panics from SQL:

1. Memory accounting - won't a panic likely leave some accounts with memory they haven't released back to their monitors, which, in turn, will lead to another panic when that monitor is... closed? I'm not sure when that closing would happen - is it when the connection is closed? Is there an issue here?
2. Descriptor leases - what if the panic is from the `LeaseManager`, which is in the sql package. Leases are shared state (as opposed to per-query), so corrupt state there is bad. Perhaps we should move all the lease stuff to a new package, or exclude some high-level method from the panics that we swallow.

I don't know... Generally, although I understand the motivation, I wouldn't have made this change. I don't think any broad category of panics are safe to recover from. As Radu said in another context, if we had a big-hammer way to "restart" a whole module without restarting the server, then that'd make it safe.

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-:-Ki5jFIivVghJZ3vm6XN:b-1dor2i)*
<!-- Sent from Reviewable.io -->

knz commented:
Your first concern is legitimate, and I will prepare an extension to recover from that.

The second concern is legitimate too, and I think the best way to address it is to move the lease manager to a separate package too.

Your general comment is sound, but the overall concern here cannot be addressed without a major refactor. Then the question becomes whether we still want to address #9884 for 1.0.
@bdarnell @petermattis what do you think?

RaduBerinde commented:
I second Andrei's concerns. We can reason about crash recovery (by talking about the "on disk"/database state), but it's very hard to reason about whether a module will still function properly after it hit and recovered from a panic.

I also fear this will have us moving toward having to write "exception"-safe code (like RAII, in our case "resource release only happens in `defer`") to support recovering from arbitrary panics (which is nightmarish IMO).

---

Review status: 0 of 8 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-:-Ki5vUJg7Jd-YqcbjfrV:bu3h2x0)*
<!-- Sent from Reviewable.io -->

knz commented:
Of course this discussion needs to be placed in the context of #9884 which highlights the poor UX of seeing the entire server crash because a client issued a bit of SQL which we don't support.

All this points into the direction of running sql connections in a separate process, like pg and TiDB are doing.

knz commented:
I would recommend continuing the discussion from here: https://github.com/cockroachdb/cockroach/issues/9884#issuecomment-295326907

RaduBerinde commented:
> seeing the entire server crash because a client issued a bit of SQL which we don't support.

This is an overstatement - we don't crash on SQL which we don't support. We crash due to bugs. I don't think we have data suggesting that panics that are due to specific SQL statements are wide-spread.

knz commented:
Most of the SQL-specific panics we've seen so far were innocuous corner cases in the implementations (i.e. programmer errors), like members of AST nodes not properly initialized, semantics not checked in the right order, etc.

knz commented:
In practice my concern is stated in https://github.com/cockroachdb/cockroach/issues/9884#issuecomment-295326907: "should we attempt to recover, at the risk of having a server in a partially broken state after a panic has been recovered, or not recover at all, at the risk of crashing one node after another as a client attempts to re-issue a faulty query from one node to another through a load balancer?"

cuongdo commented:
@bdarnell do you have thoughts on this PR?

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiW8JPFcf5aTAAJMYV3:bgzjh5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
> ```Go
> // The skip parameter indicates how many call frames to skip to
> // start the analysis.
> func isRecoverableSQLPanic(skip int) bool {
> ```

This scares me. If we're going to attempt to recover from some panics, I'd much prefer catching errors which implement a specific `Recoverable` interface. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 39 of 171 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/migrations/migrations.go, line 322 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8LwuLHjpWW4ZyNUM:-KiW8LwvkAtKegvr46vh:b6cf8ql) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/migrations/migrations.go#L322)):*
> ```Go
> 	var err error
> 	for retry := retry.Start(retry.Options{MaxRetries: 5}); retry.Next(); {
> 		res, panicErr := r.sqlExecutor.ExecuteStatements(session, alterStmt, nil)
> ```

Why is this called `panicErr` instead of `err`? In general we've found it error-prone to have `error` variables not named `err`.

---

*[pkg/sql/executor.go, line 441 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCpkBUvTwxq-dnO0e:-KiWCpkBUvTwxq-dnO0f:b-qtl88m) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L441)):*
> ```Go
> 		// is really needed.
> 		txn = client.NewTxn(e.cfg.DB)
> 		if err := txn.SetIsolation(session.DefaultIsolationLevel); err != nil {
> ```

Why did this move?

---

*[pkg/sql/executor.go, line 526 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCJRSIbQWH3mcGedx:-KiWCJRSIbQWH3mcGedy:bdux4kc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L526)):*
> ```Go
> 	defer func() {
> 		if r := recover(); r != nil {
> 			panicErr = errors.Errorf("panic while executing %q: %s", stmts, r)
> ```

Can this block be shared with the one in `Prepare`?

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiWAgKL5SbRXjKBVh5O:bqle0vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This scares me. If we're going to attempt to recover from some panics, I'd much prefer catching errors which implement a specific `Recoverable` interface. 
</blockquote></details>

But if we could panic an object with a `Recoverable` interface, we could just return an error. The goal of this change is to handle runtime-initiated errors (like nil pointer panics) more gracefully. 

---

*[pkg/sql/executor.go, line 1934 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCPX1vtM_6JoYiGXM:-KiWCPX1vtM_6JoYiGXN:biclu89) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1934)):*
> ```Go
> 		}
> 		if !strings.HasPrefix(function,
> 			"github.com/cockroachdb/cockroach/pkg/sql.(*Executor).ExecuteStatements") {
> ```

Aren't there two roots, ExecuteStatements and Prepare?

---

*[pkg/sql/executor.go, line 1942 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWBNpkyKoMH8juM5JN:-KiWBNpkyKoMH8juM5JO:b5w81un) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1942)):*
> ```Go
> 		break
> 	}
> 	// The following loop strips any additional `runtime.*` prefixes
> ```

I'm confused by the two loops here. This appears to be looking at the caller of ExecuteStatements, but I'm not sure why. The logic I would expect is a single loop, which returns false when it hits any non-runtime, non-sql package and returns true when it hits ExecuteStatements. Am I misunderstanding things?

---

*[pkg/sql/prepare.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW9lahTFn6ZCtj9MxH:-KiW9laiXvVV8TaGo9Z3:bu4v152) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/prepare.go#L77)):*
> ```Go
> 	// Prepare the query. This completes the typing of placeholders.
> 	stmt, fatalErr, err := e.Prepare(query, ps.session, placeholderHints)
> 	if err != nil || fatalErr {
> ```

It's surprising if `fatalErr` can be true with a nil `err`. What does that mean?

It doesn't look like `v3Conn.handleRecoveredPanic` does the right thing in that case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiWDzAIVQZItT6PdEwH:b-uboaz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

But if we could panic an object with a `Recoverable` interface, we could just return an error. The goal of this change is to handle runtime-initiated errors (like nil pointer panics) more gracefully. 
</blockquote></details>

Handling runtime-initiated errors and translating them into errors scares me. We're not even closing the connection when such panics occur (unless I'm completely misreading this PR).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiWMEmSRzPDlQIIo81_:b15q5ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Handling runtime-initiated errors and translating them into errors scares me. We're not even closing the connection when such panics occur (unless I'm completely misreading this PR).
</blockquote></details>

Turning runtime-initiated panics into errors is exactly the goal of this PR (and #9884). 

I'm hesitantly in favor of this change because it seems like we still have a significant number of queries of death that would be safe to recover from in this way. But I agree that it's a scary change to be making at this late date. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 10 files reviewed at latest revision, 16 unresolved discussions.

---

*[pkg/migrations/migrations.go, line 322 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8LwuLHjpWW4ZyNUM:-KiWExPzOFX7ZxOFH8P0:b-dkr6nj) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/migrations/migrations.go#L322)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why is this called `panicErr` instead of `err`? In general we've found it error-prone to have `error` variables not named `err`.
</blockquote></details>

Simplified.

---

*[pkg/sql/executor.go, line 441 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCpkBUvTwxq-dnO0e:-KiWFxHYsoFgXp3r3ZwR:bw51rrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L441)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why did this move?
</blockquote></details>

The move transforms the error case for txn initialization failure from a panic to an irrecoverable error (as it should be).

---

*[pkg/sql/executor.go, line 526 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCJRSIbQWH3mcGedx:-KiWHqMqS_veM1-CecsR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L526)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can this block be shared with the one in `Prepare`?
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiWHtSFrDuC4ZNfPIHv:bhuuuy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Handling runtime-initiated errors and translating them into errors scares me. We're not even closing the connection when such panics occur (unless I'm completely misreading this PR).
</blockquote></details>

Yes we are -- pgwire catches this case and closes the connection.

---

*[pkg/sql/executor.go, line 1934 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWCPX1vtM_6JoYiGXM:-KiWHxNDHowW1ds-kw74:b6wgciz) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1934)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Aren't there two roots, ExecuteStatements and Prepare?
</blockquote></details>

Argh, correct. Fixed.

---

*[pkg/sql/executor.go, line 1942 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiWBNpkyKoMH8juM5JN:-KiWHzCCWCDITkZidV4n:b7p1kw3) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1942)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'm confused by the two loops here. This appears to be looking at the caller of ExecuteStatements, but I'm not sure why. The logic I would expect is a single loop, which returns false when it hits any non-runtime, non-sql package and returns true when it hits ExecuteStatements. Am I misunderstanding things?
</blockquote></details>

There is a nuance there. I clarified in a comment.

---

*[pkg/sql/prepare.go, line 77 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW9lahTFn6ZCtj9MxH:-KiWRgPHoMBlCkA5nLbr:br62q9h) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/prepare.go#L77)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

It's surprising if `fatalErr` can be true with a nil `err`. What does that mean?

It doesn't look like `v3Conn.handleRecoveredPanic` does the right thing in that case.
</blockquote></details>

It can't be true - err is always set in that case. I simplified the condition accordingly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

knz commented:
@andreimatei I fixed the memory monitoring problem. Now even an emergency shutdown will do the right thing with monitors.

petermattis commented:
Review status: 0 of 10 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[pkg/sql/executor.go, line 1917 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-KiW8JPFcf5aTAAJMYV2:-KiWdz4asjVjSTCFiwqe:bs889iq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8510d10be070096df14d79da1f82b4047d0b0010/pkg/sql/executor.go#L1917)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes we are -- pgwire catches this case and closes the connection.
</blockquote></details>

Ok, I misread the PR which addresses my concern about not closing the connection. This still makes me anxious, though. The linked issue suggests having a switch to enable panic mode vs transform-panics-to-errors-and-close-connection mode.

Agreed that it feels late in the cycle to be making this change. Perhaps it should be held for early in the 1.1 cycle. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063)*
<!-- Sent from Reviewable.io -->

knz commented:
@petermattis you are right I had forgotten about the setting. Added `sql.internal_error_recovery.enabled`  for this.

Now that the setting is in, I think the discussion is really what the default should be. We could enable by default and if there is too much inconvenience "in the field" we can advise users to disable recovery. And conversely, of course.

petermattis commented:
> Now that the setting is in, I think the discussion is really what the default should be. We could enable by default and if there is too much inconvenience "in the field" we can advise users to disable recovery. And conversely, of course.

Similar to other risky changes, I think we should default to off and enable on all of our test clusters as soon as this goes in.

knz commented:
One more possible user for #15301.

bdarnell commented:
> Similar to other risky changes, I think we should default to off and enable on all of our test clusters as soon as this goes in.

I think that's backwards for this change - we want panics on our test clusters to make failures more visible, but we want to soften the impact of query-specific bugs in end-user clusters. It doesn't make sense to make this change now unless we're going to turn it on by default.

knz commented:
Just holler when y'all have decided what's best here.

petermattis commented:
> I think that's backwards for this change - we want panics on our test clusters to make failures more visible, but we want to soften the impact of query-specific bugs in end-user clusters. It doesn't make sense to make this change now unless we're going to turn it on by default.

Yes, we want to soften the impact of query-specific bugs in end-user clusters, but we don't want to introduce a worse problem while doing so. I'm still anxious about converting panics to errors and the possibility that doing so can leave the server in an inconsistent state where it corrupts user data. Disabling this feature by default and getting some experience with running it on our test clusters is one avenue to mitigating that risk.

knz commented:
Ok what about leaving the default to disabled, and if users start hitting this problem we can tell them to enable the thing. I don't foresee there will be mission-critical deployments in the first six months anyways.

bdarnell commented:
Our test clusters are not currently panicking, so we're not going to get any confidence about how this change behaves by running it there. I think that merging this in the disabled state by default is a risk with no benefit. I'd say either merge and enable it or don't merge it at all until 1.1. 

---

Review status: 0 of 10 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/server/admin.go, line 183 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-Kia3bIAWCAmSwvV4JDV:-Kia3bIAWCAmSwvV4JDW:b-5txjwo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca2efd67d13a608930c621f7e5e5a0c6af038429/pkg/server/admin.go#L183)):*
> ```Go
> 	ctx, session := s.NewContextAndSessionForRPC(ctx, args)
> 	defer session.Finish(s.server.sqlExecutor)
> 	r, panicErr := s.server.sqlExecutor.ExecuteStatements(session, "SHOW DATABASES;", nil)
> ```

My comment about `panicErr` apples to all of these too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15063#-:-Kia3X2dYuuF6aHZU_M5:btsie5c)*
<!-- Sent from Reviewable.io -->

knz commented:
"Our test clusters are not currently panicking" -> I can modify one of the loadgens to issue random queries that evaluate the new builtin `crdb_internal.internal_error()` which purposefully triggers a panic.

petermattis commented:
> "Our test clusters are not currently panicking" -> I can modify one of the loadgens to issue random queries that evaluate the new builtin crdb_internal.internal_error() which purposefully triggers a panic.

But that's somewhat uninteresting as that panic is already covered by a unit test. The panics I'm worried about are in parts of the SQL code which mutate persistent state. For example, a panic in the lease manager. 

> Our test clusters are not currently panicking, so we're not going to get any confidence about how this change behaves by running it there.

True. Hmm.

knz commented:
So what do we do here?

I could resurrect the PR and extend it so that a panic in a distsql processor is also caught to abort the session. Or I could kill the PR and have bugs in the SQL part of distsql kill entire clusters randomly. 

What do you think?

petermattis commented:
> So what do we do here?
>
> I could resurrect the PR and extend it so that a panic in a distsql processor is also caught to abort the session. Or I could kill the PR and let bugs in the SQL part of distsql kill entire clusters randomly.
>
> What do you think?

I'm very wary of trying to catch all panics out of a particular package as I doubt we're safe with regards to cleanup. I'd rather have an opt-in mechanism via an annotation of the panic that should be converted to an error. I realize this can result in a cluster being taken down by a query. I'd rather have that than trying to debug non-deterministic problems caused by the incomplete recovery of a panic.

knz commented:
Ok

What about ... egregating SQL processing into a separate set of goroutines from leases, schema updates and KV accesses. And then only catch panics in the SQL goroutine.

petermattis commented:
> What about ... segregating SQL processing into a separate set of goroutines from leases, schema updates and KV accesses. And then only catch panics in the SQL goroutine.

There will be overhead to switching goroutines. And I don't think that buys you anything over inserting defer-recover statements around particularly blocks of code (i.e. catch panics thrown by particular bits of functionality and convert them to errors).

jordanlewis commented:
Drive-by: I think adding some kind of sqlpanic() would be fairly manageable, and the initial conversion to using those wouldn't be too bad. I would also prefer a perfect (or close) automatic panic detector for sql panics, but if we can't get there I think a manual annotation wouldn't hurt too badly.

knz commented:
Thanks Jordan for keeping both feet on the ground. I propagated your comment to https://github.com/cockroachdb/cockroach/issues/9884#issuecomment-311143046 and I propose to continue the discussion about a possible design there.

I will delete this PR for now. We can discuss and/or find alternate approaches as comments to #9884.
